syncer/error.go

```diff
@@ -70,3 +75,53 @@ func originError(err error) error {
 	}
 	return err
 }
+
+// handleSpecialDDLError handles special errors for DDL execution.
+// it only ignore `invalid connection` error (timeout or other causes) for `ADD INDEX` now.
+// `invalid connection` means some data already sent to the server,
+// and we assume that the whole SQL statement has already sent to the server for this error.
+// if we have other methods to judge the DDL dispatched but timeout for executing, we can update this method.
+// NOTE: we must ensure other PK/UK exists for correctness.
+// NOTE: when we are refactoring the shard DDL algorithm, we also need to consider supporting non-blocking `ADD INDEX`.
+func (s *Syncer) handleSpecialDDLError(tctx *tcontext.Context, err error, ddls []string, index int, conn *DBConn) error {
+	// must ensure only the last statement executed failed with the `invalid connection` error
+	if len(ddls) == 0 || index != len(ddls)-1 || errors.Cause(err) != mysql.ErrInvalidConn {
```

Comment:
Have a question for the `failed SQL index`, the underlying implementation in `ExecuteSQLWithIgnoreError`, if failed in the `commit` procedure, the returned index is `len(ddls) - 1`. Such as the DDLs are ``alter table t add column col int, add index col(`col`)``, what happens to the transaction if commit failed?

Reply:
`begin` and `commit` has no effect on DDL, and DDL is always executed before `commit`.